### PR TITLE
Show how to use picolibc in a couple of demos

### PR DESCRIPTION
**Pull request by keith-packard**
Show how to use picolibc in the CORTEX_M3_MPS2_QEMU_GCC and  RISC-V_RV32_QEMU_VIRT_GCC demos

Description
-----------
This series shows how Picolibc can be used with FreeRTOS with changes to two demonstration applications. The modifications can be broken down into discrete steps:

 1. Set `configUSE_PICOLIBC_TLS = 1` when using Picolibc with TLS enabled
 2. Add TLS support to the linker script. This has no effect when not using TLS variables
 3. Replace any POSIX apis provided to support stdio and/or malloc with code suitable for picolibc
 4. Adjust CFLAGS and LDFLAGS to pass --specs=picolibc.specs in place of any other C library-specfic flags

This series includes some unrelated changes; in  CORTEX_M3_MPS2_QEMU_GCC, it appeared to intend to elide unused code with the linker but didn't include the `--gc-sections` flag to the linker, and I had trouble building the Risc-V demo because of some feature conflicts in its configuration. I can pull those out into a separate series if desired.

Requirements
-----------
This test requires an arm toolchain with both newlib and picolibc and a risc-v toolchain with picolibc. On Debian (stable, testing or unstable) or Ubuntu (22.04 or later) systems, these can be obtained with:
```
# apt install picolibc-arm-none-eabi libnewlib-arm-none-eabi picolibc-riscv64-unknown-elf
```

Test Steps
-----------
```
$ cd Demo/CORTEX_M3_MPS2_QEMU_GCC
$ make
$ size build/RTOSDemo.axf
   text	   data	    bss	    dec	    hex	filename
   15016	    240	3187744	3203000	 30dfb8	build/RTOSDemo.axf
$ make clean; make PICOLIBC=1
$ size build/RTOSDemo.axf
   text	   data	    bss	    dec	    hex	filename
   9380	    148	3187576	3197104	 30c8b0	build/RTOSDemo.axf
$ cd ../RISC-V_RV32_QEMU_VIRT_GCC
$ make -C build/gcc PICOLIBC=1
$ size build/gcc/output/RTOSDemo.elf
   text	   data	    bss	    dec	    hex	filename
  63428	    116	  89232	 152776	  254c8	build/gcc/output/RTOSDemo.elf
```

Test Requirements
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Skptak** [comment]
Hey @keith-packard first I want to thank you for raising this pull request with a new feature for FreeRTOS! 
I see that you're failing some CBMC proofs, and have a manifest.yml file issue. I wanted to ask if these issues are something you wanted to directly work on, or if you would be okay with others cloning down your fork and adding to your pull request?

**Comment by aggarg** [comment]
> Hey @keith-packard first I want to thank you for raising this pull request with a new feature for FreeRTOS! I see that you're failing some CBMC proofs, and have a manifest.yml file issue. I wanted to ask if these issues are something you wanted to directly work on, or if you would be okay with others cloning down your fork and adding to your pull request?

Both these issues should be addressed once https://github.com/FreeRTOS/FreeRTOS/pull/946 is merged.

**Comment by keith-packard** [comment]
> Both these issues should be addressed once #946 is merged.

Thanks, I was going to ask what I needed to fix as I couldn't figure out what was broken...

File: FreeRTOS/Demo/CORTEX_M3_MPS2_QEMU_GCC/syscall.c
Comment (paulbartell) [inline FreeRTOS/Demo/CORTEX_M3_MPS2_QEMU_GCC/syscall.c:82]: Note: We should probably switch these demos over to using semihosting properly for stdout. I'll give this a go in another PR.

Comment (keith-packard) [reply FreeRTOS/Demo/CORTEX_M3_MPS2_QEMU_GCC/syscall.c:82]: I figured it was simplest to leave the sample mostly alone and have it work the same with both C libraries? Might be nice to see how semihosting works in the sample though; it's a very useful tool in early SoC bring up as you don't need the clock tree set up correctly and driving a UART when working with real hardware.

Comment (paulbartell) [reply FreeRTOS/Demo/CORTEX_M3_MPS2_QEMU_GCC/syscall.c:82]: Noted. I will address this later on.

**Comment by paulbartell** [comment]
@keith-packard I added some small changes to enable newlib thread local storage as well and always enable picolibc thread local storage when picolibc is used. Does this look reasonable to you?

**Comment by keith-packard** [comment]
> @keith-packard I added some small changes to enable newlib thread local storage as well and always enable picolibc thread local storage when picolibc is used. Does this look reasonable to you?

Those additional commits seem to include a bunch of changes not mentioned in the logs and that aren't obviously related to the goals of this series?

**Comment by paulbartell** [comment]
@keith-packard With the exception of the changes to the vAssertCalled, it's mostly whitespace changes (to satisfy the auto-formatter), changes in the task stack sizes to allow Thread Local Storage to work with both picolibc and newlibc, and make the main_full.c set of "demos" run successfully (seems to not work with configUSE_TIME_SLICING  enabled).

**Comment by keith-packard** [comment]
sounds good. Do you need anything else before merging these in?

**Review by paulbartell — APPROVED** [review]

**Comment by paulbartell** [comment]
Looks good to me. I'll get a second review and we can get this PR merged in.

**Review by cobusve — APPROVED** [review]

**Review by paulbartell — APPROVED** [review]

**Pull request merged**
